common/thorhelper/thorcommon.hpp

```diff
@@ -673,6 +677,76 @@ class THORHELPER_API IndirectCodeContext : implements ICodeContext
 protected:
     ICodeContext * ctx;
 };
+
+
+
+// Wraps an IEclProcess so that when IEclProcess::perform is called, EclProcessEx::perform
+// can construct an ICodeContext that returns the workflow id.
+// (It cannot be done by wrapping IGlobalCodeContext because workflow id is known only when
+// IEclProcess::perform is called. Workflow id isn't known before that.)
+struct THORHELPER_API EclProcessEx : implements IEclProcess, public CInterface
+{
+    // Wraps ICodeContext and overrides getWorkflowId() method
+    class CodeContextEx : extends IndirectCodeContext
+    {
+        unsigned wfid;
+    public:
+        CodeContextEx(ICodeContext * _ctx, unsigned _wfid) : IndirectCodeContext(_ctx), wfid(_wfid) {}
+        virtual ~CodeContextEx() {};
+        virtual unsigned getWorkflowId() const override { return wfid; }
+    };
+    // Wraps IGlobalCodeContext and returns a constructed ICodeContext that can return workflow id
+    struct GlobalCodeContextExtra : implements IGlobalCodeContext
+    {
+        GlobalCodeContextExtra(IGlobalCodeContext * _gctx, unsigned _wfid) : gctx(_gctx)
+        {
+            codeContextEx = new CodeContextEx(gctx->queryCodeContext(), _wfid);
+        }
+        virtual ~GlobalCodeContextExtra()
+        {
+            delete codeContextEx;
+        }
+        virtual ICodeContext * queryCodeContext()
+        {
+            return codeContextEx;
+        }
+        virtual void fail(int, const char *) { gctx->fail(0, NULL); }
+
+        virtual bool isResult(const char * name, unsigned sequence) { return gctx->isResult(name, sequence); }
+        virtual unsigned getWorkflowId() const override { return gctx->getWorkflowId(); }
```

Comment:
I don't think IGlobalCodeContext::getWorkflowId() is or will be actually used anywhere?

ICodeContext::getWorkflowId() will be, but I don't see anywhere that can call IGlobalCodeContext::getWorkflowId() directly, or via generated code ?

I think it must remain untouched though (not removed), because it will break pre-existing queries if it is removed (if the VMT changes).

@ghalliday - if it is never used (I removed it and definitions as a test and can't see anywhere that generates it), and it's never needed. If we can't remove it (to avoid breaking queries), can we rename it as ```
virtual unsigned deprecatedGetWorkUnitId() // no longer used, but keeping to preserve VMT integrity 
```

?